src/communication/RpcClient.cpp

```diff
@@ -10,3 +10,371 @@
+			                         std::move(maybe_handle).value(),
+			                         std::move(expire));
+		}
+	}
+
+	return std::move(callback_handle);
+}
+
+RpcClient::InvokeHandle RpcClient::invokeMethod(
+    datamodel::builder::Payload&& payload, Callback&& callback) {
+	return invokeMethod(builder_.build(std::move(payload)),
+	                    std::move(callback));
+}
+
+RpcClient::InvokeHandle RpcClient::invokeMethod(Callback&& callback) {
+	return invokeMethod(builder_.build(), std::move(callback));
+}
+
+RpcClient::InvokeFuture RpcClient::invokeMethod(
+    datamodel::builder::Payload&& payload) {
+	// Note: functors need to be copy constructable. We work around this by
+	// wrapping the promise in a shared_ptr. Unique access to it will be
+	// assured by the implementation at the core of invokeMethod - it only
+	// allows exactly one call to the callback via std::call_once.
+	auto promise = std::make_shared<std::promise<MessageOrStatus>>();
+	auto future = promise->get_future();
+	auto handle = invokeMethod(
+	    std::move(payload), [promise](MessageOrStatus maybe_message) mutable {
+		    promise->set_value(maybe_message);
+	    });
+
+	return {std::move(future), std::move(handle)};
+}
+
+RpcClient::InvokeFuture RpcClient::invokeMethod() {
+	// Note: functors need to be copy constructable. We work around this by
+	// wrapping the promise in a shared_ptr. Unique access to it will be
+	// assured by the implementation at the core of invokeMethod - it only
+	// allows exactly one call to the callback via std::call_once.
+	auto promise = std::make_shared<std::promise<MessageOrStatus>>();
+	auto future = promise->get_future();
+	auto handle =
+	    invokeMethod([promise](MessageOrStatus maybe_message) mutable {
+		    promise->set_value(maybe_message);
+	    });
+
+	return {std::move(future), std::move(handle)};
+}
+
+RpcClient::RpcClient(RpcClient&&) = default;
+RpcClient::~RpcClient() = default;
+
+RpcClient::InvokeFuture::InvokeFuture() = default;
+RpcClient::InvokeFuture::InvokeFuture(InvokeFuture&& other) noexcept = default;
+RpcClient::InvokeFuture& RpcClient::InvokeFuture::operator=(
+    InvokeFuture&& other) noexcept = default;
+
+RpcClient::InvokeFuture::InvokeFuture(std::future<MessageOrStatus>&& future,
+                                      InvokeHandle&& handle) noexcept
+    : callback_handle_(std::move(handle)), future_(std::move(future)) {}
+
+}  // namespace uprotocol::communication
+
+///////////////////////////////////////////////////////////////////////////////
+namespace {
+namespace detail {
+
+using namespace uprotocol;
+using namespace std::chrono_literals;
+
+auto PendingRequest::operator>(const PendingRequest& other) const {
+	return when_expire > other.when_expire;
+}
+
+ScrubablePendingQueue::~ScrubablePendingQueue() {
+	const v1::UStatus cancel_reason = []() {
+		v1::UStatus reason;
+		reason.set_code(v1::UCode::CANCELLED);
+		reason.set_message("ExpireWorker shutting down");
```

Comment:
Make this something more alarming since this list should never have anything in it - could label it as a possible memory leak

Reply:
Done